dkg-gadget/src/worker.rs

```diff
@@ -478,42 +493,168 @@ where
 						.get_public_key()
 						.unwrap()
 						.get_element()
 						.serialize_uncompressed()
 						.to_vec();
-					let offchain = self.backend.offchain_storage();
 
-					if let Some(mut offchain) = offchain {
-						offchain.set(STORAGE_PREFIX, OFFCHAIN_PUBLIC_KEY, &pub_key);
-					}
+					self.gossip_public_key(pub_key, dkg_msg.round_id);
 				}
+
+				self.next_rounds = Some(next_rounds);
 			}
 		}
 
+		self.handle_public_key_broadcast(dkg_msg.clone());
+
 		self.send_outgoing_dkg_messages();
 
 		self.process_finished_rounds();
 	}
 
+	/// Offchain features
+
+	fn listen_and_clear_offchain_storage(&mut self, header: &B::Header) {
+		let at = BlockId::hash(header.hash());
+		let next_dkg_public_key = self.client.runtime_api().next_dkg_pub_key(&at).ok();
+
+		if next_dkg_public_key.is_some() {
+			let offchain = self.backend.offchain_storage();
+			if let Some(mut offchain) = offchain {
+				if offchain.get(STORAGE_PREFIX, AGGREGATED_PUBLIC_KEYS).is_some() {
+					offchain.remove(STORAGE_PREFIX, AGGREGATED_PUBLIC_KEYS);
+
+					offchain.remove(STORAGE_PREFIX, SUBMIT_KEYS_AT);
+
+					self.dkg_state.listening_for_pub_key = false;
+					self.aggregated_public_keys.keys_and_signatures = vec![];
+				}
+			}
+		}
+	}
+
+	fn gossip_public_key(&mut self, public_key: Vec<u8>, round_id: RoundId) {
+		let public = self
+			.key_store
+			.authority_id(&self.key_store.public_keys().unwrap())
+			.unwrap_or_else(|| panic!("Halp"));
+
+		if let Ok(signature) = self.key_store.sign(&public, &public_key) {
+			let encoded_signature = signature.encode();
+			let message = DKGMessage::<AuthorityId, DKGPayloadKey> {
+				id: public.clone(),
+				round_id,
+				payload: DKGMsgPayload::PublicKeyBroadcast(DKGPublicKeyMessage {
+					pub_key: public_key.clone(),
+					signature: encoded_signature.clone(),
+				}),
+			};
+
+			self.gossip_engine
+				.lock()
+				.gossip_message(dkg_topic::<B>(), message.encode(), true);
+
+			self.aggregated_public_keys
+				.keys_and_signatures
+				.push((public_key.clone(), encoded_signature));
+			info!(target: "dkg", "queued_keygen: gossiping local node  {:?} public key and signature", public)
+		}
+		else {
+			error!(target: "dkg", "queued_keygen: Could not sign public key");
+		}
+	}
+
+	fn handle_public_key_broadcast(&mut self, dkg_msg: DKGMessage<Public, DKGPayloadKey>) {
+		if !self.dkg_state.listening_for_pub_key {
+			return
+		}
+
+		match dkg_msg.payload {
+			DKGMsgPayload::PublicKeyBroadcast(msg) => {
+				let key_and_sig = (msg.pub_key, msg.signature);
+				if !self.aggregated_public_keys.keys_and_signatures.contains(&key_and_sig) {
+					self.aggregated_public_keys.keys_and_signatures.push(key_and_sig);
+
+					if let Some(latest_header) = self.latest_header.as_ref() {
+						let threshold = self.get_threshold(latest_header).unwrap() as usize;
+						let num_authorities = self.queued_validator_set.authorities.len();
+						let threshold_buffer = num_authorities.saturating_sub(threshold) / 2;
```

Comment:
can you describe what this buffer is and why its there?

Reply:
I wanted to aggregate a little more public keys above the threshold value,